worlds/timespinner/Options.py

```diff
@@ -53,6 +53,100 @@ class BossRando(Choice):
     option_unscaled = 2
     alias_true = 1
 
+class BossRandoType(Choice):
+    """
+    Sets what type of boss shuffling occurs.
+    Shuffle: Bosses will be shuffled amongst each other
+    Chaos: Bosses will be randomized with the chance of duplicate bosses
+    Singularity: All bosses will be replaced with a single boss
+    Manual: Bosses will be placed according to the Boss Rando Overrides setting
+    """
+    display_name = "Boss Randomization Type"
+    option_shuffle = 0
+    option_chaos = 1
+    option_singularity = 2
+    option_manual = 3
+
+def boss_option(boss: str) -> Dict[Optional, Or]:
+    return {
+        Optional(boss): Or(
+            And({
+                Optional("FelineSentry"): And(int, lambda n: n >= 0),
+                Optional("Varndagroth"): And(int, lambda n: n >= 0),
+                Optional("AzureQueen"): And(int, lambda n: n >= 0),
+                Optional("GoldenIdol"): And(int, lambda n: n >= 0),
+                Optional("Aelana"): And(int, lambda n: n >= 0),
+                Optional("Maw"): And(int, lambda n: n >= 0),
+                Optional("Cantoran"): And(int, lambda n: n >= 0),
+                Optional("Genza"): And(int, lambda n: n >= 0),
+                Optional("Nuvius"): And(int, lambda n: n >= 0),
+                Optional("Vol"): And(int, lambda n: n >= 0),
+                Optional("Prince"): And(int, lambda n: n >= 0),
+                Optional("Xarion"): And(int, lambda n: n >= 0),
+                Optional("Ravenlord"): And(int, lambda n: n >= 0),
+                Optional("Ifrit"): And(int, lambda n: n >= 0),
+                Optional("Sandman"): And(int, lambda n: n >= 0),
+                Optional("Nightmare"): And(int, lambda n: n >= 0)
+            }, lambda d: any(v > 0 for v in d.values())),
+            "FelineSentry",
+            "Varndagroth",
+            "AzureQueen",
+            "GoldenIdol",
+            "Aelana",
+            "Maw",
+            "Cantoran",
+            "Genza",
+            "Nuvius",
+            "Vol",
+            "Prince",
+            "Xarion",
+            "Ravenlord",
+            "Ifrit",
+            "Sandman",
+            "Nightmare")
+    }
+
+class BossRandoOverrides(OptionDict):
+    """Manual mapping of bosses to the boss they will be replaced with. Bosses that you don't specify will be the
+    vanilla boss."""
+    schema = Schema({
+        **boss_option("FelineSentry"),
+        **boss_option("Varndagroth"),
+        **boss_option("AzureQueen"),
+        **boss_option("GoldenIdol"),
+        **boss_option("Aelana"),
+        **boss_option("Maw"),
+        **boss_option("Cantoran"),
+        **boss_option("Genza"),
+        **boss_option("Nuvius"),
+        **boss_option("Vol"),
+        **boss_option("Prince"),
+        **boss_option("Xarion"),
+        **boss_option("Ravenlord"),
+        **boss_option("Ifrit"),
+        **boss_option("Sandman"),
+        **boss_option("Nightmare")
```

Comment:
Uhhh, does this not just work as

```py
Optional(Or("FelineSentry", "Varndagroth", "AzureQueen", ...):
```

?

The "Or" in schema is not exclusive or, it allows multiple keys

```py

schema = Schema({Or("a", "b"): 1})
schema.validate({"a": 1})
>>> {"a": 1}
schema.validate({"a": 1, "b": 1})
>>> {"a": 1, "b": 1}
schema.validate({"a": 1, "b": 1, "c": 1})
>>> ... SchemaValidationError ...
```

Reply:
```py
from schema import Schema, Optional, And, Or

bosses = [
    "FelineSentry",
    "Varndagroth",
    "AzureQueen",
    "GoldenIdol",
    "Aelana",
    "Maw",
    "Cantoran",
    "Genza",
    "Nuvius",
    "Vol",
    "Prince",
    "Xarion",
    "Ravenlord",
    "Ifrit",
    "Sandman",
    "Nightmare",
]

schema = Schema({
    Optional(Or(*bosses)): Or(
        And({
            Optional(boss): And(int, lambda n: n >= 0) for boss in bosses
        }, lambda d: any(v > 0 for v in d.values())),
        *bosses)
    }
)

schema.validate({
    "FelineSentry": "Maw",
    "AzureQueen": {"Ravenlord": 50, "Ifrit": 50},
    "Genza": {"Nightmare": 100},
})
```

This passes (and is in my opinion a much nicer way overall to write this)

Reply:
ok I'll make that change. Thanks for that I'm not super familiar with what all you can do with Schema